### DISCUSSION TITLE
CustomControls ALPHA now available!

### DISCUSSION THREAD

**Discussion by WaynePhillipsEA** [twinbasic/twinbasic#467] [General]
After what seems like a very long wait(!), CustomControls are now available in twinBASIC.  This is an experimental feature and as such is currently only available to those that have pre-ordered their twinBASIC licence.    Instructions will be sent out to all registered customers over the next few hours.

In the current alpha release, the form designer does not support attaching code to forms/controls and forms themselves cannot be instantiated outside of the form designer.  To launch a form into a real Windows form (for testing mouse events for example), press the big red "PREVIEW" button in the form designer.

There is a new Sample project (Sample 6) that contains 9 example forms for you to play with.  Here are a couple more videos to whet your appetite.   Excluding blank lines, the grid control is less than 500 lines of BASIC code, and supports things like column resizing and scrolling.  It is of course not quite finished yet 😉

Initial documentation will be available over the next few days.  In the meantime, if you have any questions, please just ask away here.

![CustomControlsDemo_Grid](https://user-images.githubusercontent.com/25738109/137507487-6c7a44da-edb7-4ed4-9328-15ee2ac68be1.gif)
![CustomControls_Slider](https://user-images.githubusercontent.com/25738109/137507513-69da203d-7a0d-454d-a0cc-05690dbab5ae.gif)

Have a nice weekend folks

**Comment by WaynePhillipsEA** [comment]
All licence keys have now been issued via email to those that have pre-ordered twinBASIC.   If you think you've been missed, please first check your email SPAM folder, and if still missing please get in touch and we will look into it for you.

**Comment by uwbwsvd** [comment]
After a short look. I see this. (firewall give an alarm).
There is no progress.
![grafik](https://user-images.githubusercontent.com/41230018/137525160-2ffa5b59-ba47-4a88-baba-3cf1e7cca857.png)

**Reply by uwbwsvd** [reply]
This is working:

- Example_DateTimeControl.tbform
- Example_TextRendering.tbform

-

**Reply by WaynePhillipsEA** [reply]
The form designer creates a new websocket connection to the compiler, and this is what might trigger a firewall / Defender warning.  You must allow that connection and then hopefully it should work (but might require you to close the .tbform and restart VS code).

**Reply by Kr00l** [reply]
> The form designer creates a new websocket connection to the compiler, and this is what might trigger a firewall / Defender warning. You must allow that connection and then hopefully it should work (but might require you to close the .tbform and restart VS code).

I have the same issue. I allowed it and then the form opened. Then it said license key not found within the form. After closure and reopening I have the same waiting cycles.

**Reply by WaynePhillipsEA** [reply]
Apologies.  v0.11.1 is now released.  Hopefully fixes this issue!

**Reply by WaynePhillipsEA** [reply]
Note: the compiler will still report v0.11.0 as I didn't rebuild the compiler EXE for this quickfix.   The version number should be 0.11.1 listed in the Extension panel.

**Reply by WaynePhillipsEA** [reply]
Assuming this fix works, I'll put in a more permanent fix tomorrow.

**Reply by uwbwsvd** [reply]
This is working:
- Example_DateTimeControl.tbform
- Example_TextRendering.tbform 
are working
the rest not.

I deleted the sample workspace and create a new one with 0.11.1

**Reply by WaynePhillipsEA** [reply]
v0.11.14 is now available with a few further tweaks.  When opening the form designer you will now see a log above the loading animation.  If it still gets stuck at that point, please screenshot the log and post it here.  Thanks.

**Reply by uwbwsvd** [reply]
![grafik](https://user-images.githubusercontent.com/41230018/137573990-145df62b-def8-4fb3-824d-8b974e758416.png)

**Reply by WaynePhillipsEA** [reply]
Thanks @uwbwsvd.   What OS are you running?

**Reply by uwbwsvd** [reply]
![grafik](https://user-images.githubusercontent.com/41230018/137574472-6002ba7e-0b3d-4b72-9ce7-cd7ab26f68bc.png)

**Reply by uwbwsvd** [reply]
I want to remember: this is working
![grafik](https://user-images.githubusercontent.com/41230018/137574618-2c4438f4-c0cd-4a25-9ae8-2eddfdc96349.png)

**Reply by WaynePhillipsEA** [reply]
@uwbwsvd yes, that's the very strange part!

**Reply by WaynePhillipsEA** [reply]
So Example_Slider.tbform doesn't work for you, right?   If so, try adding a slider control from the toolbox onto your Example_TextRendering.tbform.  Does that work?

**Reply by WaynePhillipsEA** [reply]
v0.11.16 is now available that includes a little bit more detail in the on-screen log, plus it adds a button at the end of the log so that you can continue to open the form when the loading is stuck (which might help us identify what is wrong).

**Reply by uwbwsvd** [reply]
![grafik](https://user-images.githubusercontent.com/41230018/137577193-09ec16d8-8951-4630-8a40-a03ea00cd21a.png)

**Reply by uwbwsvd** [reply]
![grafik](https://user-images.githubusercontent.com/41230018/137577220-80ea1a02-7c09-4b19-8ff9-f8e3147c93d4.png)

**Reply by WaynePhillipsEA** [reply]
Thanks!   Could we see another screenshot after clicking the ignore and continue button?

**Reply by uwbwsvd** [reply]
![grafik](https://user-images.githubusercontent.com/41230018/137578154-ded17163-8b74-4166-9b0e-80de069dd671.png)

**Reply by WaynePhillipsEA** [reply]
Intriguing.  If I could reproduce/debug it, I'm sure this would be fixed very quickly!.   I'm just publishing v0.11.37 that may help.  Let me know if you get any different behaviour with it.

**Reply by uwbwsvd** [reply]
**First: All examples are running now!**
Here I see this:
![grafik](https://user-images.githubusercontent.com/41230018/137580146-4b1b41a7-7d4a-405a-9759-f41af9fbb4fe.png)

**Reply by WaynePhillipsEA** [reply]
Great, thanks for persevering with this.  So we've now identified the problem area, but we need to look further for the source of the problem.  We should have this fixed properly in the next 24 hours.  When you see this error, some controls on the form won't be editable with the property sheet, until the bug is fixed.

**Reply by WaynePhillipsEA** [reply]
OK so I think this is fixed in v0.11.41 being published now.   Localization strikes again... the new JSON serializer in tB was outputting floating point values with regional settings (e.g. potentially '1,234' rather than '1.234' as per the JSON spec).  Fingers crossed!

**Reply by uwbwsvd** [reply]
I would say: :+1:

**Reply by WaynePhillipsEA** [reply]
Excellent, thanks for confirming!

**Reply by uwbwsvd** [reply]
Closing VS-Code with left open designer and reopen it the form-designer would fail starting. Closing the form-designer and reopen it everything works fine.

**Reply by WaynePhillipsEA** [reply]
Yes, I've got that one noted, thanks.

**Reply by WaynePhillipsEA** [reply]
@Kr00l just checking is it working for you now?

**Reply by Kr00l** [reply]
> @Kr00l just checking is it working for you now?

yes. thx

**Comment by mansellan** [comment]
Just downloaded it and gave it a try.

Hugely impressive stuff I have to say! It's clear that this approach is going to work, and it's now just a case of filling in the missing functionality and building out the basic controls.

One thing that would be nice is to minimise the amount of JSON that is needed for serialisation. I see why the _classId and _className properties are there, but it would be nice if there were a way to avoid them by mapping the classes from the property names in the JSON somehow. Also, it would be optimal if only modified properties were stored - perhaps serialise the "default" instance of a control, then the actual instance, then store only the diffs. Minimising the JSON like this would be a big help down the line when needing to conduct Pull Requests, for example.

It also occurs to me that the JSON could be used to apply styling themes - after the control has been deserialised, a style could be applied by deserialising a theme (i.e. a subset of properties selected by the theme author) on top of the instance.

That's for the future though, it's looking great so far!

**Reply by WaynePhillipsEA** [reply]
Thanks @mansellan glad you like it!

I'll have a think about the JSON serialisation, as it does seem a bit verbose at the moment.  As Mike mentioned in one of the earlier discussions, Access does similarly to what you propose with storing defaults at the top of the file before the main serialisation.

Thanks for trying it, and for your feedback!

**Comment by WaynePhillipsEA** [comment]
There is now some initial documentation up for implementing CustomControls:
https://github.com/WaynePhillipsEA/twinbasic/wiki

It's very much a work-in-progress... but it's a start.

**Comment by uwbwsvd** [comment]
Expected behavior?
The grid wents beneath the first button. The following buttons are but above.  The buttons lost ther width property.

https://user-images.githubusercontent.com/41230018/137773892-3ff869fa-4aa7-4eb2-99b4-7ab0d26ca1d5.mp4

**Reply by WaynePhillipsEA** [reply]
Thanks @uwbwsvd yes that looks slightly off.  I'll take a look at it.

**Comment by wopag** [comment]
Looking good, look forward to running code on events :) 

Small issue (you probably know) right click a control and it shows 2 popup menus, the built in cut/paste popup and behind that your cut/paste/send back/front

**Reply by WaynePhillipsEA** [reply]
Thanks!

The right-click popup menu is a bit of an issue, as VS code doesn't currently allow to remove or customize the built-in one.  There is a bodge fix in the tB extension that tries to move the focus away from the form designer which causes the built-in menu to close, leaving only our custom menu.  It relies on timing to work though, and so is somewhat unreliable.   I need to ask the VS Code team to accept a PR for disabling the built-in menu.

**Comment by WaynePhillipsEA** [comment]
v0.11.561 is now available and we've got OnGotFocus, OnLostFocus, OnKeyDown and OnKeyUp events for each element.  We've also got ElementTabIndex and ElementTabStop properties for each element, which allows an individual control like the Grid to allow focusing and tabbing between each of its internal elements (cells).

The form designer and form engine now support TabIndex (Long) and TabStop (Boolean) properties if you expose them on your control.   The examples in Sample 6 have all been updated to support focused controls and tabbing, and so I would advise checking them out to see how the new events work in practice.   The grid example now supports tabbing between cells (forwards and backwards), and also supports arrow-key navigation.    I'll update the documentation in a few days.

Description attributes on exposed fields and properties will now show up in the form designer, so you can design controls with proper help text for each field.  Again, look at Sample 6, as all exposed fields in the examples now have some help text.

The Slider examples have also been reworked.   You might notice that some of the Slider examples look like scrollbars.  There's a reason for that... which will become apparent next week.

**Comment by dmrvb** [comment]
I really should have asked sooner, but please please can we have mouseEnter and mouseLeave events? The absence of these in VB6 has caused me many major and time consuming problems..
In the slightly longer term, I make significant use of control arrays. This capability in tB would be really good. As these are new controls I am not too fussed about 100% compatibility, I am quite happy to rework my code to make best use of tB

**Reply by WaynePhillipsEA** [reply]
Thanks @dmrvb,  OnMouseEnter and OnMouseLeave are already available.  Here's the current list of events that any element within a custom control can listen for:

```
        OnClick
        OnDblClick
        OnMouseDown
        OnMouseUp
        OnMouseEnter
        OnMouseLeave
        OnMouseMove
        OnScrollH
        OnScrollV
        OnGotFocus
        OnLostFocus
        OnKeyDown
        OnKeyUp
```
I am considering control arrays for the CustomControls, but I suspect they can wait for a little later down the line.

**Comment by WaynePhillipsEA** [comment]
> On Saturday, 23 October 2021, EduardoVB ***@***.***> wrote:
>     OnClick
>     OnDblClick
>     OnMouseDown
>
> Just curious, why are you naming the events preceded by "On" and not just
> the traditional "Click", "MouseDown", etc. that VB always used?

These are just the internal events corresponding to the elements within
your custom control.  The 'on' prefix just helps clarify which of the
ElementDescriptor fields are actually events.

You can name the real events exposed by your custom control as you see
fit.  Hope that makes sense.

**Reply by EduardoVB** [reply]
Yes, I figured that inmediately and deleted the question. Thank you.

**Reply by WaynePhillipsEA** [reply]
No worries, it was a question others will no doubt ask, so I figured it was worth keeping.  Cheers.

**Comment by WaynePhillipsEA** [comment]
v0.11.1331 is now available, with the main addition being a functioning Textbox control:

![twinBASIC TextBox Example](https://user-images.githubusercontent.com/25738109/139521559-979e69e6-ef7d-45b2-adda-90fe6a7667a9.gif)

WaynesTextbox control supports most of the standard Windows features such as cut/copy/paste, select-all, word selection and movement via arrow keys etc.  Only about 500 lines of BASIC code.  Some work is still required here as the hotkeys are hard coded for english at the moment (e.g. Ctrl-A for select-all, Ctrl-C for copy etc).  RTL languages are not yet supported, and still lots of tests to write before this is finished.

All twinBASIC control elements can now support a feature called TextDecorations which allow adding squiggles/underlines and changing background/foreground fill colors for parts of the text rendering.  See WaynesTextbox Paint event for examples.   You simply create a TextDecorations object and call the AddSquiggle / AddUnderline / AddBackgroundFill / ChangeTextFill members, passing in the start index of the character, and the length of text to apply it to.    In future we will also support font modification via TextDecorations, which will open up possibilities for a rich-textbox.

**Reply by WaynePhillipsEA** [reply]
There's currently some bugs around surrogate pairs that will be fixed in the next update.

**Reply by WaynePhillipsEA** [reply]
Unicode and surrogate pair issues have now been resolved in the example textbox control, as of v0.11.1355, so you can now enter your emojis

**Comment by wopag** [comment]
I know you have your timeline roadmap so just a personal comment,  it seems secondary to get custom controls to work instead of getting an interface to work with off the shelf .COM controls?  like active3d and codejock etc ?  Personally wont ever use a custom control.

We could start testing the code events callbacks etc ?

**Reply by WaynePhillipsEA** [reply]
A lot of the custom controls work is to be re-used for native controls (and ActiveX containers).  It is important to get a solid structure and framework in place that can support all of the features we need going forward, in particular for the form designer, JSON serialization, form engine, events and compiler integration.  These things are now coming together nicely and will make things so much easier when we move on to the native controls.     64-bit support is top of the list though before we get on to the native controls.

**Reply by wopag** [reply]
Knew you had a plan, Looking good

**Reply by mwolfe02** [reply]
> ... (and ActiveX containers) ...

Can I infer from this comment that there will be support for creating custom ActiveX controls in twinBASIC that can be added to forms in Microsoft Access (and other Office apps)?  That could be a real game-changer.

**Reply by WaynePhillipsEA** [reply]
@mwolfe02 yes, that will definitely be happening.

**Reply by bclothier** [reply]
Just as a follow-up regarding the ActiveX design, do you anticipate making an element a window, or would you be able to let us determine what will be a window and what won't be? I imagine that in the case of datasheet grid and the like, we might not necessarily want every cell be a window in some situations (at least they don't do this in other Microsoft products). Not sure if VB6 gave you that level of control when creating an ActiveX control.

**Comment by wopag** [comment]
Right click is working better - but find if you press and hold the right mouse button down for 3-4 seconds only vcode menu appears from then on.  TB menu never refreshes (stays on screen at its last position)

**Reply by WaynePhillipsEA** [reply]
I haven't changed anything in this regard yet.   It's just a bit fragile.

**Reply by WaynePhillipsEA** [reply]
I've done some work on the context menus in v0.11.1368.  They should now be working much better / more reliably.

**Comment by WaynePhillipsEA** [comment]
Getting there....

![realForm1](https://user-images.githubusercontent.com/25738109/139716727-fbc4930e-6d09-444c-a448-c035389af592.gif)

**Comment by WaynePhillipsEA** [comment]
As of v0.12.1, the custom controls feature is now unlocked for all users (not needing a licence key).  It is still an ALPHA feature, so be aware that there certainly will be bugs at this stage!

Sample 6 (Custom Controls) has been updated to offer a simple walkthrough of the sample forms.  As you will discover, real forms can now be created along with real event bindings.    

You can also build an EXE from your project, though be aware that the built size of the executable is currently more bloated than it will be once CustomControls support has been more refined.   Sample 6, with 10 forms currently compiles to ~2MB in size.   The Sample 6 EXE has been tested back to Windows XP without issue.

Further documentation to follow.    Have fun!

**Comment by WaynePhillipsEA** [comment]
In v0.12.117 I worked on getting the bloated compiled EXE issue under control.   For example, the Sample 6 CustomControls containing 10 forms and 8 custom controls now compiles down to a 733KB EXE, down from the 2.2MB size in the v0.12.1 release.  

Considering that the compiled EXE contains the whole forms engine, plus the form designs and code, **and** the custom control implementations, 733KB is not half bad.
